nova/scheduler/filters/shard_filter.py

```diff
@@ -148,18 +254,28 @@ def host_passes(self, host_state, spec_obj):
         if self._ALL_SHARDS in shards:
             LOG.debug('project enabled for all shards %(project_shards)s.',
                       {'project_shards': shards})
-            return True
         elif host_shard_names & set(shards):
             LOG.debug('%(host_state)s shard %(host_shard)s found in project '
                       'shards %(project_shards)s.',
                       {'host_state': host_state,
                        'host_shard': host_shard_names,
                        'project_shards': shards})
-            return True
         else:
             LOG.debug('%(host_state)s shard %(host_shard)s not found in '
                       'project shards %(project_shards)s.',
                       {'host_state': host_state,
                        'host_shard': host_shard_names,
                        'project_shards': shards})
             return False
+
+        if k8s_shard:
+            matches = any(host_shard == k8s_shard
+                          for host_shard in host_shard_names)
+            if not matches:
```

Comment:
Couldn't we use `if k8s_shard in host_shard_names` here? Would be much shorter.
We'd need to explicitly `return False` and `return True` then, but I don't see that as a downside.
As an alternative, we could do `matches = k8s_shard in host_shard_names`, too.